pkg/chain/chain.go

```diff
@@ -1,164 +1,239 @@
+// 	BalanceMonitor() (BalanceMonitor, error)
+// }
+
+// Handle represents a handle to an implementation of operator functions
+// for a particular chain.
 type Handle interface {
-	// Address returns client's ethereum address.
-	Address() common.Address // TODO: use implementation-agnostic type
+	// Name returns the name of this chain handle. The name should be
+	// descriptive of the chain this handle represents; for example, for
+	// Ethereum it might return "Ethereum", for Celo "Celo". It can optionally
+	// include additional information if it would be useful in logs or debug
+	// output; for example, Ethereum handles might include network id
+	// information.
+	Name() string
+	// OperatorID returns the operator id associated with this handle.
+	OperatorID() OperatorID
+
+	// chain.BaseHandle from keep-core
 	// StakeMonitor returns a stake monitor.
 	StakeMonitor() (chain.StakeMonitor, error)
 	// BalanceMonitor returns a balance monitor.
 	BalanceMonitor() (chain.BalanceMonitor, error)
 	// BlockCounter returns a block counter.
-	BlockCounter() chain.BlockCounter
+	BlockCounter() (chain.BlockCounter, error)
 	// Signing returns a signer interface allowing to sign and verify messages
 	// using the chain implementation-specific mechanism as well as to
 	// convert between public key and address.
 	Signing() chain.Signing
-	// BlockTimestamp returns given block's timestamp.
-	// In case the block is not yet mined, an error should be returned.
-	BlockTimestamp(blockNumber *big.Int) (uint64, error)
 
-	BondedECDSAKeepFactory
-	BondedECDSAKeep
+	PublicKeyToOperatorID(publicKey *cecdsa.PublicKey) OperatorID
+
+	// BondedECDSAKeepManager returns a handle to the on-chain component used to
+	// interact with bonded ECDSA keeps.
+	BondedECDSAKeepManager() (BondedECDSAKeepManager, error)
 }
 
-// BondedECDSAKeepFactory is an interface that provides ability to interact with
-// BondedECDSAKeepFactory ethereum contracts.
-type BondedECDSAKeepFactory interface {
-	// RegisterAsMemberCandidate registers client as a candidate to be selected
-	// to a keep.
-	RegisterAsMemberCandidate(application common.Address) error
+// BondedECDSAKeepManager is an interface whose implementations allow for
+// managing, monitoring, and creating bonded ECDSA keeps as well as the
+// operator's eligibility for participation in such keeps.
+type BondedECDSAKeepManager interface {
+	// TBTCApplicationHandle returns a handle for interacting with the tBTC
+	// application associated with this BondedECDSAKeepManager. Returns nil with
+	// an error if no tBTC application exists for this manager.
+	TBTCApplicationHandle() (TBTCHandle, error)
 
 	// OnBondedECDSAKeepCreated installs a callback that is invoked when an
 	// on-chain notification of a new bonded ECDSA keep creation is seen.
 	OnBondedECDSAKeepCreated(
 		handler func(event *BondedECDSAKeepCreatedEvent),
 	) subscription.EventSubscription
 
+	// GetKeepCount returns number of keeps.
```

Comment:
I know the scope of this PR is a higher level, but leaving the comment here for the future. As this is no longer a function on `BondedECDSAKeepFactory` interface I'd mention in the comment that it's a number of keeps created by the associated factory.

Reply:
One of the things I'm interested in thoughts on that I didn't call out explicitly is whether it makes sense to call this `BondedECDSAKeepManager` instead of sticking to `BondedECDSAKeepFactory`.

Reply:
At first, I found the rename controversial. But if the idea of this diverges from the idea of reflecting the `BondedECDSAKeepFactory` contract very closely it makes sense to rename it.

Reply:
It may or may not right now, but I think we don't want these to be attached to each other. The underlying concept being that `chain` is the interface, but it dictates as little as possible about the underlying implementation for a specific supported chain.